lib/jsonapi/deserializable/resource.rb

```diff
@@ -1,110 +1,155 @@
-require 'jsonapi/deserializable/resource_dsl'
+require 'jsonapi/deserializable/resource/configuration'
+require 'jsonapi/deserializable/resource/dsl'
+require 'jsonapi/parser/resource'
 
 module JSONAPI
   module Deserializable
     class Resource
-      include ResourceDSL
+      extend DSL
 
       class << self
         attr_accessor :type_block, :id_block, :attr_blocks,
-                      :has_one_rel_blocks, :has_many_rel_blocks
+                      :has_one_rel_blocks, :has_many_rel_blocks,
+                      :configuration
       end
 
-      self.attr_blocks = {}
+      @class_cache = {}
+
+      self.configuration       = Configuration.new
+      self.attr_blocks         = {}
       self.has_one_rel_blocks  = {}
       self.has_many_rel_blocks = {}
 
       def self.inherited(klass)
         super
-        klass.type_block  = type_block
-        klass.id_block    = id_block
-        klass.attr_blocks = attr_blocks.dup
+        klass.configuration       = configuration.dup
+        klass.type_block          = type_block
+        klass.id_block            = id_block
+        klass.attr_blocks         = attr_blocks.dup
         klass.has_one_rel_blocks  = has_one_rel_blocks.dup
         klass.has_many_rel_blocks = has_many_rel_blocks.dup
       end
 
+      def self.configure
+        yield(configuration)
+      end
+
+      def self.[](name)
+        @class_cache[name] ||= Class.new(self)
+      end
+
       def self.call(payload)
         new(payload).to_h
       end
 
       def initialize(payload)
+        Parser::Resource.parse!(payload)
         @document = payload
         @data = @document['data']
         @type = @data['type']
         @id   = @data['id']
         @attributes    = @data['attributes'] || {}
         @relationships = @data['relationships'] || {}
         deserialize!
+        freeze
       end
 
       def to_hash
         @hash
       end
       alias to_h to_hash
 
+      attr_reader :reverse_mapping
+
       private
 
-      def deserialize!
-        @hash = {}
-        deserialize_type!
-        deserialize_id!
-        deserialize_attrs!
-        deserialize_rels!
+      def configuration
+        self.class.configuration
       end
 
-      def deserialize_type!
-        return unless @type && self.class.type_block
-        instance_exec(@type, &self.class.type_block)
+      def register_mappings(keys, path)
+        keys.each do |k|
+          @reverse_mapping[k] = path
+        end
       end
 
-      def deserialize_id!
-        return unless @id && self.class.id_block
-        instance_exec(@id, &self.class.id_block)
+      def deserialize!
+        @reverse_mapping = {}
```

Comment:
Seems like an add place for this, not sure what the purpose is and not seeing a spec for it (might be missing something)